scripts directory README documentation (#1437)
raft-tech/TANF-app: scripts/README.md

# scripts

The TANF app uses several scripts through its lifecycle. 
These don't all get used by or interacted with by us too often,
but some are mission critical during deployment and review of 
the application. When a developer is working on these scripts, 

> they should update this documentation so future developer's 

Reviewer: ```suggestion
they should update this documentation so future developers 
```

> understand the role of the scripts

Reviewer 2: ```suggestion
understand the role of the scripts.
```

# set-backend-env-vars.sh 

## usage 

```bash

./scripts/set-backend-env-vars.sh CGAPPNAME_BACKEND CG_SPACE
```

## args

> `CGAPPNAME_BACKEND` the sub domain of the app you are trying to set up environment variables for.

Reviewer: Might be useful to provide some real world examples - ie `tdp-backend-raft`, `tanf-dev`, etc

Reviewer 2: Yes, agreed, please provide a concrete example. Could you also please link to our other documentation which lists out all the spaces and backends? Like [this one](https://github.com/raft-tech/TANF-app/blob/c0c9423dcd4d9b87930eb655a74dd8f2701e3dcf/docs/Technical-Documentation/TDP-environments-README.md) for the backends.

`CG_SPACE` the space the app you are trying to set up for exists in.

## description

Determine the appropriate BASE_URL for the deployed instance based on the
provided Cloud.gov App Name
Use Shell Parameter Expansion to replace localhost in the URL
Dynamically generate a new `DJANGO_SECRET_KEY`
Dynamically set `DJANGO_CONFIGURATION` based on Cloud.gov Space

# sudo-check.sh

## usage

```
./scripts/sudo-check.sh

```

## Args

no args

## description

> check if the `sudo` command is present, installs it if it isn't.

Reviewer 2: Please provide the reader context that these "check" scripts are used by our CircleCI pipeline and that they are ran inside our buildpack containers. Additionally, please use formal grammar and full-length sentences in descriptions.

# cf-checks.sh

## usage

```bash
./scripts/cf-check.sh
```

## description

check if the cf command is present, if its not, install it and all of its dependencies.

## Args

no args

# copy-login-gov-keypair.sh

## usage

```bash

> ./scripts/copy-login-gov-keypair.sh

Reviewer 2: How are the args listed below supplied to the script? Are they command line arguments or environment variables?

```

## Description

Copies Login.gov JWT_KEY + JWT_CERT from one Cloud.gov application to another.

## Args

`SOURCE_APP` The app to copy keys from.
`DEST_APP` The app to copy keys to.

# deploy-backend.sh

> 

Reviewer 2: Missing usage?

## description

The deployment strategy you wish to employ ( rolling update or setting up a new environment)

> DEPLOY_STRATEGY=${1}

Reviewer: Would be helpful to explain each deploy strategy and how they differ.

The application name  defined via the manifest yml for the frontend
CGHOSTNAME_BACKEND=${2}

# deploy-frontend.sh

## Args

```bash

# The deployment strategy you wish to employ ( rolling update or setting up a new environment)
DEPLOY_STRATEGY=${1}

# The application name  defined via the manifest yml for the frontend
CGHOSTNAME_FRONTEND=${2}
```

# deploy-infrastructure-dev.sh

> 

Reviewer 2: Also missing usage?

> Requires installation of jq - https://stedolan.github.io/jq/download/

Reviewer 2: Is this entirely missing description area or was there a change of formatting style?

## Args

no args

# deploy-infrastructure-staging.sh

Requires installation of jq - https://stedolan.github.io/jq/download/

## Args

no args

> # docker-check.sh

Reviewer 2: Let's move this up next to the other "check" scripts.

## Usage

```bash
./scripts/docker-check.sh
```
## Args

no args

## Description

Used during ci/cd to check if docker is installed in the environment. 
It installs it if it is not.

> # docker-compose-check.sh

Reviewer 2: Move this as well next to the "checks" scripts above.

Used during ci/cd to check if docker compose is installed in the environment. 
It installs it if it is not.

## Args

no args

# git-secrets-check.sh

> ensure we have correct configs in place

Reviewer: ```suggestion
ensure that no secrets have been committed to the repo
```

grep will return non-zero code if nothing found, failing the build

## Args

no args

# localstack-setup.sh

> 

Reviewer 2: Missing usage? I thought this had arguments or are they supplied via .env?

Author: They are from env vars

Author: do we want to cover ENV vars in the readme seperately @abottoms-coder ?

Reviewer 2: > do we want to cover ENV vars in the readme seperately @abottoms-coder ?

I want the arguments for a given script to be explained in the section for that script like you did in `set-backend-env-vars.sh` section. If they're from the environment, please specify that. Please let me know if you have any questions.

Create the bucket used by the Django app
Enable object versioning on the bucket

## Args

no args

# trufflehog-check.sh

## Usage

```bash
./scripts/trufflehog-check.sh <branch-target>
```

## Description

> Install truffleHog in a venv

Reviewer 2: ```suggestion
Install truffleHog in a python virtual environment
```

Additionally, let's combine all the below sentence fragments into cohesive prose.

Get the hash of the latest commit in the target branch.
Look at all commits since the last merge into raft-tdp-main
Entropy checks on large git diffs
if there are issues, they will be listed then script will abort

## Args

`branch-target` the branch name you want to check.

# codecov-check.sh
  
## Usage

```bash
./scripts/codecov-check.sh
```

## Args

no args

## Description

  Import Codecov PGP public keys
  Download codecov uploader
  Download SHA signatures and validate installed codeov
  Validate successful installation of codecov

# zap-hook.py

## Usage

> This script is used in an arguement passed inside of `zap-scanner.sh`. It is not used directly

Reviewer: ```suggestion
This script is used in an argument passed inside of `zap-scanner.sh`. It is not used directly
```

## Args

no args

## description

Python hook that can be used to disable ignored rules in ZAP scans.
This hook runs after the ZAP API has been successfully started.

This is needed to disable passive scanning rules which we have set to IGNORE

> in the ZAP configuration files. Due to an unresolved issue with the scripts

Reviewer 2: ```suggestion
in the [ZAP configuration file](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/reports/zap.conf). Due to an unresolved issue with the scripts
```

### Where it's used
This script is no longer used by other automation as it has been obsolesced by our Terraform flows.

## deploy-infrastructure-staging.sh
### Usage
```
scripts/deploy-infrastructure-staging.sh
```

### Arguments

no args

### Description
Script runs our CircleCI job `deploy-infrastructure-staging` using your CloudFoundry login credentials; it expects that you had run `cf login --sso` prior. 
Requires installation of jq - https://stedolan.github.io/jq/download/

### Where it's used
This script is no longer used by other automation as it has been obsolesced by our Terraform flows.

# "Check" scripts
## sudo-check.sh

<details>
<summary>Details</summary>
    
### usage
---
```
./scripts/sudo-check.sh
```
### Arguments
no args
### Description
This script installs the `sudo` command and all of its dependencies if it is not already present.

### Where it's used
---
</details>

## cf-checks.sh
<details>
<summary>Details</summary>
    
### usage
---
```bash
./scripts/cf-check.sh
```
---
### Arguments
no args

### Description
This script installs the CloudFoundry `cf` command and all of its dependencies if it is not already present.

> ### Where it's used 

Reviewer 2: Nit, this is empty

    
</details>

## docker-check.sh

<details>
<summary>Details</summary>
    
### Usage
---
```bash
./scripts/docker-check.sh
```
---
    
### Arguments
no args

### Description
This script installs the docker ecosytem and all of its dependencies if it is not already present. This is used by our CircleCI CI/CD pipelines.

### Where it's used 
    
</details>
    
## docker-compose-check.sh

<details>
<summary>Details</summary>
    
### Usage
---
```
./scripts/docker-compose-check.sh
```
---
    
### Arguments
no args

### Description
This script installs the `docker-compose` command and all of its dependencies if it is not already present. This is used by our CircleCI CI/CD pipelines.
### Where it's used 

>     

Reviewer 2: Nit, this is empty

</details>

## git-secrets-check.sh
<details>
<summary>Details</summary>
    
### Usage
---
```
./scripts/git-secrets-check.sh
```
---
### Arguments

no args

### Description

This script ensures that no secrets have been committed to the TANF repo. We leverage [Awslab's git-secrets tool](https://github.com/awslabs/git-secrets.git) to scan code to be uploaded. Developers can set up a pre-commit hook locally so this scan will run before committing/pushing by checking the README on their github repo linked earlier.

> ### Where it's used 

Reviewer 2: Nit, this is empty

    
</details>

## trufflehog-check.sh

<details>
<summary>Details</summary>
    
### Usage
---
```bash
./scripts/trufflehog-check.sh <branch-target>
```
---
### Arguments

`branch-target` the branch name you want to check.

### Description

Installs truffleHog in a python virtual environment and gets the hash of the latest commit in the target branch.
Looks at all commits since the last merge into raft-tdp-main, and entropy checks on large git diffs. 
If there are issues, they will be listed then script will abort.

### Where it's used 

>     

Reviewer 2: Nit, this is empty

</details>

## codecov-check.sh
  
<details>
<summary>Details</summary>
    
### Usage
---
```bash
./scripts/codecov-check.sh
```
---
### Arguments
no args

### Description

Check if code cov is installed, and if it isn't, the script installs it, and checks the integrity of the binary.

### Where it's used 

>     

Reviewer 2: Nit, this is empty

  * `tanf-staging` uses the `Staging` configuration
  * All other spaces use the `Development` configuration
  * This refers to the Django settings class that will be used within the `tdpservice.settings.cloudgov` module. 

### Where it's used
The script `deploy-backend` invokes this script if the `DEPLOY_STRATEGY` is `initial`, `bind` or `rebuild`. Can optionally be run locally on its own.

## [copy-login-gov-keypair.sh](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/scripts/copy-login-gov-keypair.sh)

### Usage

```bash
./scripts/copy-login-gov-keypair.sh <source-app> <dest-app>
```

### Arguments

* `SOURCE_APP`: The app to copy keys from.
  * Example: `tdp-backend-raft`
* `DEST_APP`: The app to copy keys to.
  * Example: `tdp-backend-qasp`

### Description

Copies Login.gov JWT_KEY + JWT_CERT from one Cloud.gov application to another. This assumes that the user is already logged in to the Cloud Foundry CLI via the `cf login --sso` command. Both apps must exist within the same space.

### Where it's used

This script is only a convenience tool for developers, it has no direct usage in CI.

## deploy-backend.sh

### Usage

> ```

Reviewer 3: ```suggestion
```bash
```

### Description
Mostly for use in our CircleCI pipelines, this script ensures the desired codebase is deployed into a given Cloud.gov application space.

### Where it's used

`deploy-frontend.sh` is used in [CircleCI config](../.circleci/config.yml)'s job `deploy-frontend` which is used by `deploy-cloud-dot-gov` command. This script is primarily via in CircleCI but with CloudFoundry set up and logged in locally, this can also be used by a developer manually. Please also see our [CircleCI documentation](../docs/Technical-Documentation/circle-ci.md#deploy-frontend).

## deploy-infrastructure-dev.sh

### Usage

```
scripts/deploy-infrastructure-dev.sh
```

### Arguments

#### CF Service Keys

 CF_USERNAME_DEV
 CF_PASSWORD_DEV

### Description
This script will manually run CircleCI's job `deploy-infrastructure-dev` which deploys the default Terraform codebase to a desired Cloud.gov application name in the tanf-dev space. Please see [Terraform docs](../terraform/README.md) for more information.

Requires local client to have CloudFoundry and CircleCI CLI tools to be configured and logged in.

Requires installation of jq - https://stedolan.github.io/jq/download/

### Where it's used

> This script is no longer used by other automation as it has been obsolesced by our Terraform flows.

Reviewer 3: above we say that its in use and here it sounds like maybe it isnt? can we confirm which is the case? assuming it is still in-use, can we reframe here? @riatzukiza @abottoms-coder

Author: I think @jorgegonzalez Could clarify this a little better than I could. Let me get  back to you on this if he doesn't. Sorry I didn't see it before the break.

Author: I think basicly, this is a dev tool more than an ops tool. I think maybe it saying it is no longer used by any automation is misleading, I don't think this was ever used by automation.

## deploy-infrastructure-staging.sh
### Usage
```
scripts/deploy-infrastructure-staging.sh
```

### Arguments

no args

### Description
Script runs our CircleCI job `deploy-infrastructure-staging` using your CloudFoundry login credentials; it expects that you had run `cf login --sso` prior. 
Requires installation of jq - https://stedolan.github.io/jq/download/

### Where it's used
This script is no longer used by other automation as it has been obsolesced by our Terraform flows.

# "Check" scripts
## sudo-check.sh

<details>
<summary>Details</summary>
    
### usage
---
```
./scripts/sudo-check.sh
```
### Arguments
no args
### Description
This script installs the `sudo` command and all of its dependencies if it is not already present.

### Where it's used
sudo-check.sh is used in the sudo-check [CircleCI command](../.circleci/config.yml#L85)

> ---

Reviewer 3: ```suggestion
```

### Where it's used 
git-secrets-check.sh is used in the [secrets-check command](../.circleci/config.yml#L329).
</details>

## trufflehog-check.sh

<details>
<summary>Details</summary>
    
### Usage
---
```bash
./scripts/trufflehog-check.sh <branch-target>
```
---
### Arguments

`branch-target` the branch name you want to check.

### Description

Installs truffleHog in a python virtual environment and gets the hash of the latest commit in the target branch.
Looks at all commits since the last merge into raft-tdp-main, and entropy checks on large git diffs. 
If there are issues, they will be listed then script will abort.

### Where it's used 
trufflehog-check.sh is also used in the [secrets-check command](../.circleci/config.yml#L329).
</details>

## codecov-check.sh
  
<details>
<summary>Details</summary>
    
### Usage
---
```bash
./scripts/codecov-check.sh
```
---
### Arguments
no args

### Description

Check if code cov is installed, and if it isn't, the script installs it, and checks the integrity of the binary.

### Where it's used 

> codecov-check.sh is used in CircleCI 'codecov-upload" [job](../.circleci/config.yml#L91).

Reviewer 3: ```suggestion
codecov-check.sh is used in CircleCI `codecov-upload` [job](../.circleci/config.yml#L91).
```

</details>

> ## localstack-setup.sh

Reviewer 3: ```suggestion
## localstack-setup.sh

<details>
<summary>Details</summary>
```

### Usage
```
./scripts/localstack-setup.sh <bucket> <region>
```

### Arguments

bucket : Name of the localstack bucket you want to create
region : Name of the localstack region you want your new bucket to be placed in

### Description
Create the S3 bucket used by the Django app
Enable object versioning on the bucket

### Where it's used
This is used in `../tdrs-backend/docker-compose.yml` to setup your local docker connections to AWS.

>     

Reviewer 3: ```suggestion
</details>

```
